### PR TITLE
Support building and testing PostgreSQL-16

### PR DESCRIPTION
**Pull request by phracek**
This pull request enables building and testing PostgreSQL-16.

The first commit modifies sources to enable building and testing.
The second commit adds dist-gen generated source for postgresql-16

### CODE REVIEW

**Comment by phracek** [comment]
This PR is blocked by #547

**Comment by zmiklank** [comment]
[test]

**Comment by hhorak** [comment]
In order to have PostgreSQL 16 in Fedora, we need to use Fedora 39 as the base, it seems that Fedora 38 is used currently and there is no PostgreSQL 16 there. Plus, we need to use `postgresql16-server` as the packages to be installed on Fedora 39.

**Comment by phracek** [comment]
Fedora failure is known:
```
Error:
 Problem: package pgaudit-1.7.0-5.fc39.x86_64 from fedora requires postgresql-server(:MODULE_COMPAT_15), but none of the providers can be installed
  - package postgresql-server-15.4-1.fc39.i686 from fedora requires postgresql(x86-32) = 15.4-1.fc39, but none of the providers can be installed
  - package postgresql-server-15.4-1.fc39.x86_64 from fedora requires postgresql(x86-64) = 15.4-1.fc39, but none of the providers can be installed
  - package postgresql16-16.1-1.fc39.x86_64 from updates conflicts with postgresql provided by postgresql-15.4-1.fc39.i686 from fedora
  - package postgresql16-16.1-1.fc39.x86_64 from updates conflicts with postgresql provided by postgresql-15.4-1.fc39.x86_64 from fedora
  - package postgresql16-contrib-16.1-1.fc39.x86_64 from updates requires postgresql16(x86-64) = 16.1-1.fc39, but none of the providers can be installed
  - conflicting requests
  - 
  - ```

**Comment by phracek** [comment]
> In order to have PostgreSQL 16 in Fedora, we need to use Fedora 39 as the base, it seems that Fedora 38 is used currently and there is no PostgreSQL 16 there. Plus, we need to use `postgresql16-server` as the packages to be installed on Fedora 39.

@hhorak This commit should fix it https://github.com/sclorg/postgresql-container/pull/548/commits/d855c5622ee5379d4d95f3d1477a1aa54b3d50de. And generated source is here: https://github.com/sclorg/postgresql-container/pull/548/commits/1fb430addca722f17a99c3074f66ca0718eac14c

[test]

**Comment by phracek** [comment]
The corresponding bugzilla is: https://bugzilla.redhat.com/show_bug.cgi?id=2255013

**Comment by zmiklank** [comment]
[test]

File: specs/multispec.yml [hunk 1]
```diff
@@ -156,6 +156,15 @@ specs:
       c9s_image_name: "sclorg/postgresql-{{ spec.short }}-c9s"
       latest_fedora: "f38"
 
+    "16":
+      version: "16"
+      prev_version: "13"
```
Comment (fila43) [inline specs/multispec.yml:161]: From my POW here should be 15

File: specs/multispec.yml [hunk 2]
```diff
@@ -156,6 +156,15 @@ specs:
...
+      short: "16"
+      prev_short: "13"
```
Comment (fila43) [inline specs/multispec.yml:163]: Same as above, there should be 15.

**Comment by fila43** [comment]
Once this [update](https://bodhi.fedoraproject.org/updates/FEDORA-2024-f55ff878b8) is pushed, Fedora test should pass

> Fedora failure is known:
> 
> ```
> Error:
>  Problem: package pgaudit-1.7.0-5.fc39.x86_64 from fedora requires postgresql-server(:MODULE_COMPAT_15), but none of the providers can be installed
>   - package postgresql-server-15.4-1.fc39.i686 from fedora requires postgresql(x86-32) = 15.4-1.fc39, but none of the providers can be installed
>   - package postgresql-server-15.4-1.fc39.x86_64 from fedora requires postgresql(x86-64) = 15.4-1.fc39, but none of the providers can be installed
>   - package postgresql16-16.1-1.fc39.x86_64 from updates conflicts with postgresql provided by postgresql-15.4-1.fc39.i686 from fedora
>   - package postgresql16-16.1-1.fc39.x86_64 from updates conflicts with postgresql provided by postgresql-15.4-1.fc39.x86_64 from fedora
>   - package postgresql16-contrib-16.1-1.fc39.x86_64 from updates requires postgresql16(x86-64) = 16.1-1.fc39, but none of the providers can be installed
>   - conflicting requests
>   - 
>   - ```
> ```

File: specs/multispec.yml (hunk 2 above)
Comment (phracek) [reply specs/multispec.yml]: Fixed by https://github.com/sclorg/postgresql-container/pull/548/commits/5fa4da9c64afffa0d29b3e4e4ca5a6cad541d5c7

File: specs/multispec.yml (hunk 1 above)
Comment (phracek) [reply specs/multispec.yml]: Fixed by https://github.com/sclorg/postgresql-container/pull/548/commits/5fa4da9c64afffa0d29b3e4e4ca5a6cad541d5c7

**Comment by phracek** [comment]
[test]

**Comment by zmiklank** [comment]
[test]

File: src/Dockerfile.fedora [hunk 3]
```diff
@@ -44,12 +48,18 @@ COPY root/usr/libexec/fix-permissions /usr/libexec/fix-permissions
 # This image must forever use UID 26 for postgres user so our volumes are
 # safe in the future. This should *never* change, the last test is there
 # to make sure of that.
+{% if spec.version == "16" %}
```
Comment (zmiklank) [inline src/Dockerfile.fedora:51]: This way of distinguishing between modularized and de-modularized versions is not really extensible. We will have  postgresql 17 in half a year, then new version every year and having so many conditions will be error-prone.

In demodularized fedoras we could install postgresql{% spec.version %}, etc.

Comment (fila43) [reply src/Dockerfile.fedora:51]: I would keep this format, at least for PG16. The new concept of packaging postgresql is entirely new, and it needs to be tested, so there could be slight changes in the feature.

**Comment by phracek** [comment]
[test]

**Comment by phracek** [comment]
Another Fedora update for postgresql-pgaudit https://bodhi.fedoraproject.org/updates/FEDORA-2024-e86c64a60b

**Comment by phracek** [comment]
Fedora fix is already in stable. Thanks @fila43 . Let's try to re-run tests.

[test]

**Comment by zmiklank** [comment]
Patch like this is needed (use postgresql16-pgaudit, instead of pgaudit for postgresql 16 in fedora):
```
diff --git a/src/Dockerfile.fedora b/src/Dockerfile.fedora
index 806621e..ad8e03e 100644
--- a/src/Dockerfile.fedora
+++ b/src/Dockerfile.fedora
@@ -54,9 +54,12 @@ RUN INSTALL_PKGS="rsync tar gettext bind-utils postgresql16-server postgresql16-
 RUN INSTALL_PKGS="rsync tar gettext bind-utils postgresql-server postgresql-contrib nss_wrapper " && \
 {% endif %}
     INSTALL_PKGS+="findutils xz" && \
-{% if spec.version not in ["9.6", "10", "11"] %}
+{% if spec.version not in ["9.6", "10", "11", "16"] %}
     INSTALL_PKGS+=" pgaudit" && \
 {% endif %}
+{% if spec.version == "16" %}
+    INSTALL_PKGS+=" postgresql16-pgaudit" && \
+{% endif %}
 {% if spec.version not in ["16"] %}
     dnf -y module enable postgresql:{{ spec.version }} && \
 {% endif %}
 ```

**Comment by phracek** [comment]
[test]

**Review by zmiklank — APPROVED** [review]

**Pull request merged**